mobile/evidence/ios/Features/Tests/LoginTests/LoginTests.swift

```diff
@@ -0,0 +1,120 @@
+        let autheUser = try? AuthenticatedLoginManager().getAuthenticationUser()
+        XCTAssertNotNil(autheUser, "The user should exist")
+    }
+
+    func testGetAuthenticationUserFailure() {
+        let autheUser = try? FailureAuthenticationLoginManager().getAuthenticationUser()
+//        viewModel.getAuthenticationUser()
+        XCTAssertNil(autheUser, "Shouldn't exist")
+    }
+
+    ///LoginEmailViewModel()
+    func testCloseButtonTapped() {
+        let viewModel = LoginEmailViewModel()
+        var closeButtonCalled = false
+        viewModel.delegateCloseButtonTapped = {
+            closeButtonCalled = true
+        }
+        viewModel.closeButtonTapped()
+        XCTAssertTrue(closeButtonCalled, "The delegate should be called")
+    }
+
+    func clearEmailInputTapped() {
+        let viewModel = LoginEmailViewModel()
+        viewModel.clearEmailInputTapped()
+        XCTAssertTrue(viewModel.emailInput.isEmpty, "Should be empty")
+        XCTAssertFalse(viewModel.isNextButtonPressed,  "Should set false")
+    }
+
+    func inputEmailTapped() {
+        let viewModel = LoginEmailViewModel()
+        viewModel.clearEmailInputTapped()
+        XCTAssertFalse(viewModel.isNextButtonPressed, "Should set false")
+    }
+
+    func testButtonNextTappedValidEmail() {
+        let viewModel = LoginEmailViewModel()
+        viewModel.emailInput = "email@valid.com"
+        viewModel.buttonNextTapped()
+        XCTAssertTrue(viewModel.isValidEmail, "Should be a valid email")
+        XCTAssertTrue(viewModel.isNextButtonPressed, "Should set True")
+        XCTAssertNotNil(viewModel.loginCheckViewModel, "Should create an instance of LoginEmailViewModel()" )
+    }
+
+
+    func testButtonNextTappedNotValidEmail() {
+        let viewModel = LoginEmailViewModel()
+        viewModel.emailInput = "invalid-email"
+        viewModel.buttonNextTapped()
+        XCTAssertFalse(viewModel.isValidEmail, "Should not be a valid email")
+        XCTAssertTrue(viewModel.isNextButtonPressed, "Should set True")
+        XCTAssertNil(viewModel.loginCheckViewModel, "Should not create an instance of LoginEmailViewModel()")
+    }
+
+    func testErrorMessageNoEmailProvided() {
+        let viewModel = LoginEmailViewModel()
+        viewModel.isNextButtonPressed = true
+        viewModel.emailInput = ""
+        let result = viewModel.errorMessage()
+        XCTAssertEqual(result, "No email provided.")
+    }
+
+    func testErrorMessageInvalidEmail() {
+        let viewModel = LoginEmailViewModel()
+        viewModel.isNextButtonPressed = true
+        viewModel.emailInput = "invalid"
+        let result = viewModel.errorMessage()
+        XCTAssertEqual(result, "That doesn't look like a valid email address!")
+    }
+
+    func testErrorMessageValidEmail() {
+        let viewModel = LoginEmailViewModel()
+        viewModel.isNextButtonPressed = true
+        viewModel.emailInput = "email@valid.com"
+        viewModel.isValidEmail = true
```

Comment:
Essa property `isValidEmail` não deveria ser assignable do lado de fora ne? Isso abre uma brecha pra alguém fazer isso la na view e "burlar" as regras de validação. Daria pra transformar ela em `private(set)`?

Reply:
Interesting. In this case, the data it's accessible but just for reading.

Reply:
However, this variable `emailInput` is responsible for collecting the data from the `TextFieldEmail`, if I put it private(set) my code brokes.

Reply:
After I add the access control modifier to the variable `isValidEmail` my test `testErrorMessageValidEmail()`  stopped work bc inside this function I need to set the variable as `true`, then to solve this I did some `gambiarra`.  Since I can't set a value to this variable I called the `buttonNextTapped` method to turn true the `isValidEmail ` variable. 

```swift

``` func testErrorMessageValidEmail() {
        let viewModel = LoginEmailViewModel()
        viewModel.emailInput = "email@valid.com"
        viewModel.buttonNextTapped()
        let result = viewModel.errorMessage()
        XCTAssertTrue(viewModel.isValidEmail, "Shoude set true")
        XCTAssertNil(result, "Should not return any message")
    }
    
    what do think about my solution?